Packet++/header/GvcpLayer.h

```diff
@@ -0,0 +1,626 @@
+		DataOverrun = 0x800D,
+		InvalidHeader = 0x800E,
+		WrongConfig = 0x800F,  // deprecated
+		PacketNotYetAvailable = 0x8010,
+		PacketAndPrevRemovedFromMemory = 0x8011,
+		PacketRemovedFromMemory = 0x8012,
+		NoRefTime = 0x8013,                     // GEV 2.0
+		PacketTemporarilyUnavailable = 0x8014,  // GEV 2.0
+		Overflow = 0x8015,                      // GEV 2.0
+		ActionLate = 0x8016,                    // GEV 2.0
+		LeaderTrailerOverflow = 0x8017,         // GEV 2.1
+		Error = 0x8FFF,
+		Unknown = 0xFFFF
+	};
+
+	std::ostream& operator<<(std::ostream& os, GvcpResponseStatus status);
+
+#pragma pack(push, 1)
+	/// @brief Gvcp request header
+	/// @note refer to the spec "15.1 Request Header". The data is stored as big-endian.
+	struct GvcpRequestHeader
+	{
+		friend class GvcpRequestLayer;
+
+	protected:
+		uint8_t magicNumber = internal::kGvcpMagicNumber;  // always fixed
+		uint8_t flag = 0;  // 0-3 bits are specified by each command, 4-6 bits are reserved, 7 bit is acknowledge
+		uint16_t command = 0;
+		uint16_t dataSize = 0;
+		uint16_t requestId = 0;
+
+	public:
+		// ------------- methods --------------
+		GvcpRequestHeader() = default;
+
+		GvcpRequestHeader(GvcpFlag flag, GvcpCommand command, uint16_t dataSize, uint16_t requestId)
+		    : flag(flag), command(hostToNet16(static_cast<uint16_t>(command))), dataSize(hostToNet16(dataSize)),
+		      requestId(hostToNet16(requestId))
+		{}
+
+		GvcpFlag getFlag() const
+		{
+			return flag;
+		}
+
+		GvcpCommand getCommand() const
+		{
+			return static_cast<GvcpCommand>(netToHost16(command));
+		}
+
+		uint16_t getDataSize() const
+		{
+			return netToHost16(dataSize);
+		}
+
+		uint16_t getRequestId() const
+		{
+			return netToHost16(requestId);
+		}
+
+		/**
+		 * @brief Verify the magic number
+		 * @return true The magic number is valid
+		 */
+		bool verifyMagicNumber() const
+		{
+			return magicNumber == internal::kGvcpMagicNumber;
+		}
+
+		/**
+		 * @brief Check if the acknowledge is required
+		 * @return true The acknowledge is required
+		 */
+		bool hasAcknowledgeFlag() const
+		{
+			constexpr GvcpFlag kAcknowledgeFlag = 0b0000001;
+			return (flag & kAcknowledgeFlag) == kAcknowledgeFlag;
+		}
+	};
+	static_assert(sizeof(GvcpRequestHeader) == internal::kGvcpRequestHeaderLength,
+	              "Gvcp request header size should be 8 bytes");
+
+	/// @brief Gvcp discovery request
+	struct GvcpDiscoveryRequest : public GvcpRequestHeader
+	{
+		// no addition fields
+
+		// ------------- methods --------------
+
+		/**
+		 * @brief Check if the broadcast is allowed
+		 * @return true The broadcast acknowledge is allowed
+		 */
+		bool hasAllowBroadcastFlag() const
+		{
+			constexpr GvcpFlag kAllowBroadcastFlag = 0b0001000;
+			return (flag & kAllowBroadcastFlag) == kAllowBroadcastFlag;
+		}
+	};
+
+	/// @brief Gvcp acknowledge header
+	/// @note refer to the spec "15.2 Acknowledge Header". The data is stored as big-endian.
+	struct GvcpAckHeader
```

Comment:
I'm still learning the GVCP protocol, but shouldn't this struct inherit from `GvcpRequestHeader`? 
If not, isn't there a common header for all GVCP packets?

Reply:
The request header is different from the acknowledge header, so I separated them.

Reply:
Is there a common header for all GVCP packets?

Reply:
Unfortunately, no.

The request (or called command):
![image](https://github.com/user-attachments/assets/47e3e39e-e291-488c-a5cc-b4a278953056)

The response (or called acknowledge):
![image](https://github.com/user-attachments/assets/a710aece-dd14-4aaa-a8f0-0781907579b1)